README.md
# Confire #
[![Build Status][build_status_img]][build_status_page] [![PyPi version][pypi_version_img]][pypi_version] [![PyPi downloads][pypi_downloads_img]][pypi_downloads] [![Stories in Ready][waffle_img]][waffle_status]

**A simple app configuration scheme using YAML and class based defaults.**

[![Confire][confire.jpg]][confire.jpg]

Confire is a simple but powerful configuration scheme that builds on the configuration parsers of Scapy, elasticsearch, Django and others. The basic scheme is to have a configuration search path that looks for YAML files in standard locations. The search path is hierarchical (meaning that system configurations are overloaded by user configurations, etc). These YAML files are then added to a default, class-based configuration management scheme that allows for easy development.

Full documentation can be found here: [http://confire.readthedocs.org/](http://confire.readthedocs.org/)

## Features ##

* Configuration files in YAML
* Hierarchical configuration search
* Class based application defaults
* Settings pulled in from the environment

## Example Usage ##

Create a file called "myapp.yaml" and place it in one of the following places:

* /etc/myapp.yaml
* $HOME/.myapp.yaml
* conf/myapp.yaml

Create some configuration values inside the file like so:

    ## Set application environment
    debug:   True
    testing: False

    ## A simple database configuration
    database:
        name: mydb
        host: localhost
        port: 5432
        user: postgres

In your code, create a file called "config.py" and add the following:

    from confire import Configuration
    from confire import environ_setting

    class DatabaseConfiguration(Configuration):

        host = "localhost"
        port = 5432
        name = "mydb"
        user = "postgres"
        password = environ_setting("DATABASE_PASSWORD", required=False)

    class MyAppConfiguration(Configuration):

        CONF_PATHS = [
            '/etc/myapp.yaml',
            os.path.expanduser('~/.myapp.yaml'),
            os.path.abspath('conf/myapp.yaml')
        ]

        debug    = False
        testing  = True
        database = DatabaseConfiguration()

    settings = MyAppConfiguration.load()

Now, everywhere in your code that you would like to access these settings values, simply use as follows:

    from config import settings

    debug = settings.get('DEBUG') or settings['DEBUG']

Voila! A complete configuration system for your application!

### A note on environment variables ###

Confire is setup to enable the use of environment variables, especially if you use the helper `environ_setting` method that comes with Confire. Just note, however, that a settings file will always override the Environment, not the other way around! Sensitive values should never be placed in a settings file anyway, but rather added to the environment at run time.

The `environ_setting` alows you to specify required or optional settings to fetch from the ENVIRONMENT as defaults. It also allows you to specify a default if the ENVVAR is not there. See the documentation for more details.

## Setup ##

The easiest and usual way to install confire is to use pip:

    pip install confire

To install the package from source, download the latests package tarball, unzip in a temporary directory and run the following command:

    python setup.py install

As always, I highly recommend the use of a virtual environment to better manage the software dependencies for your particular code base.

## About ##
There are many configuration packages available on PyPI - it seems that everyone has a different way of doing it. However, this is my prefered way, and I found that after I copy and pasted this code into more than 3 projects that it was time to add it as a dependency via PyPI. The configuration builds on what I've learned/done in configuring Scapy, elasticsearch, and Django - and builds on these principles:

1. Configuration _should not_ be Python (sorry Django). It's too easy to screw stuff up, and anyway, you don't want to deal with importing a settings file from /etc!
2. Configuration should be on a per-system basis. This means that there should be an /etc/app.yaml configuration file as well as a $HOME/.app.yaml configuration file that overwrites the system defaults for a particular user. For development purposes there should also be a $(pwd)/app.yaml file so that you don't have to sprinkle things throughout the system if not needed.
3. Developers should be able to have reasonable defaults already written in code if no YAML file has been provided. These defaults should be added in an API like way that is class based and modularized.
4. Accessing settings from the code should be easy.

So there you have it, with these things in mind I wrote confire and I hope you enjoy it!

### Contributing ###

Confire is open source, and I would be happy to have you contribute! You can contribute in the following ways:

1. Create a pull request in Github: [https://github.com/bbengfort/confire](https://github.com/bbengfort/confire)
2. Add issues or bugs on the bugtracker: [https://github.com/bbengfort/confire/issues](https://github.com/bbengfort/confire/issues)
3. Checkout the current dev board on waffle.io: [https://waffle.io/bbengfort/confire](https://waffle.io/bbengfort/confire)

You can contact me on Twitter if needed: [@bbengfort](https://twitter.com/bbengfort)

### Name Origin ###
<big>con &middot; fit</big><br />
/kôNˈfē/<br/>
*noun* duck or other meat cooked slowly in its own fat.

Origin<br />
[French] *confire*: to prepare<br \>
Also refers to the culinary art of pickling

I like cooking, and the thought of preparation in French culinary language appealed to me. The way I got here was to simply change the "g" in config to a "t". A definition lookup and boom, a name!

## Changelog ##

The release versions that are sent to the Python package index are also tagged in Github. You can see the tags through the Github web application and download the tarball of the version you'd like. Additionally PyPI will host the various releases of confire.

The versioning uses a three part version system, "a.b.c" - "a" represents a major release that may not be backwards compatible. "b" is incremented on minor releases that may contain extra features, but are backwards compatible. "c" releases are bugfixes or other micro changes that developers should feel free to immediately update to.

### v0.2.0 released on 31 July 2014 ###

* **tag**: v0.2.0
* **deployment**: July 31, 2014
* **commit**: (latest)

This release will add some new features including support for environmental variables as settings defaults, ConfigurationMissing Warnings and ImproperlyConfigured errors that you can raise in your own code to warn developers about the state of configuration.

This release also greatly increased the amount of available documentation for Confire.

### v0.1.1 released on 24 July 2014 ###

* **tag**: v0.1.1
* **deployment**: July 24, 2014
* **commit**: bdc0488

Added Python 3.3 support thanks to [@tyrannosaurus](https://github.com/tyrannosaurus) who contributed to the changes that would ensure this support for the future. I also added Python 3.3 travis testing and some other minor changes.

### v0.1.0 released on 20 July 2014 ###

* **tag**: v0.1.0
* **deployment**: July 20, 2014
* **comit**: 213aa5e

Initial deployment of the confire library.


<!-- References -->
[build_status_img]: https://travis-ci.org/bbengfort/confire.svg?branch=master
[build_status_page]: https://travis-ci.org/bbengfort/confire
[confire.jpg]: http://upload.wikimedia.org/wikipedia/commons/d/d4/Picholines_et_Olives_Nyons.jpg
[pypi_version]: https://crate.io/packages/confire/
[pypi_version_img]: https://pypip.in/v/confire/badge.png
[pypi_downloads]: https://crate.io/packages/confire/
[pypi_downloads_img]: https://pypip.in/d/confire/badge.png
[waffle_img]: https://badge.waffle.io/bbengfort/confire.png?label=ready&title=Ready
[waffle_status]: https://waffle.io/bbengfort/confire


docs/changelog.rst
.. Confire documentation file, created by Benjamin Bengfort on Wed Jul 30 20:13:44 2014.

Changelog
=========

The release versions that are sent to the Python package index are also tagged in Github. You can see the tags through the Github web application and download the tarball of the version you'd like. Additionally PyPI will host the various releases of confire.

The versioning uses a three part version system, "a.b.c" - "a" represents a major release that may not be backwards compatible. "b" is incremented on minor releases that may contain extra features, but are backwards compatible. "c" releases are bugfixes or other micro changes that developers should feel free to immediately update to.

Contributors
------------

I'd like to personally thank the following people for contributing to confire and making it a success!

- `@tyrannosaurus <https://github.com/tyrannosaurus>`_
- `@murphsp1 <https://github.com/murphsp1>`_
- `@keshavmagge <https://github.com/keshavmagge>`_
- `@ojedatony1616 <https://github.com/ojedatony1616>`_

Versions
--------

The following lists the various versions of confire and important details about them.

v0.2.0
~~~~~~

* **tag**: v0.2.0
* **deployment**: July 31, 2014
* **commit**: (latest)

This release added some new features including support for environmental variables as settings defaults, ConfigurationMissing Warnings and ImproperlyConfigured errors that you can raise in your own code to warn developers about the state of configuration.

This release also greatly increased the amount of available documentation for Confire.

v0.1.1
~~~~~~

* **tag**: v0.1.1
* **deployment**: July 24, 2014
* **commit**: bdc0488

Added Python 3.3 support thanks to `@tyrannosaurus <https://github.com/tyrannosaurus>`_ who contributed to the changes that would ensure this support for the future. I also added Python 3.3 travis testing and some other minor changes.

v0.1.0
~~~~~~

* **tag**: v0.1.0
* **deployment**: July 20, 2014
* **comit**: 213aa5e

Initial deployment of the confire library.



tests/__init__.py
# tests
# Testing for the confire module
#
# Author:   Benjamin Bengfort <benjamin@bengfort.com>
# Created:  Sun Jul 20 09:42:26 2014 -0400
#
# Copyright (C) 2014 Bengfort.com
# For license information, see LICENSE.txt
#
# ID: __init__.py [] benjamin@bengfort.com $

"""
Testing for the confire module
"""

##########################################################################
## Imports
##########################################################################

import unittest

##########################################################################
## Test Cases
##########################################################################

class InitializationTest(unittest.TestCase):

    def test_initialization(self):
        """
        Tests a simple world fact by asserting that 10*10 is 100.
        """
        self.assertEqual(10*10, 100)

    def test_import(self):
        """
        Can import confire
        """
        try:
            import confire
        except ImportError:
            self.fail("Unable to import the confire module!")

    def test_version(self):
        """
        Assert that the version is sane
        """
        import confire
        self.assertEqual("0.2.0", confire.get_version())


tests/conf_tests.py
# tests.conf_tests
# Testing the configuration module for Confire
#
# Author:   Benjamin Bengfort <benjamin@bengfort.com>
# Created:  Sun Jul 20 09:43:33 2014 -0400
#
# Copyright (C) 2014 Bengfort.com
# For license information, see LICENSE.txt
#
# ID: conf_tests.py [] benjamin@bengfort.com $

"""
Testing the configuration module for Confire

TODO: Write test for None values in config
TODO: Ensure that "empty" values e.g. [], and {} override on config
TODO: Ensure that configure method is working correctly
TODO: Ensure that testing is not dependent on a user's configuration
"""

##########################################################################
## Imports
##########################################################################

import os
import yaml
import unittest
import tempfile

from copy import copy
from confire.config import *

##########################################################################
## Environment setting on Import
##########################################################################

os.environ['TESTING_CONFIRE_PASSWORD'] = 'password'

##########################################################################
## Configuration Unit Tests
##########################################################################

class ConfigurationTests(unittest.TestCase):

    FIXTURE = {
        "myprop": "Allen",
        "mysetting": False,
        "items": ["apples", "bananas", "oranges"],
        "nested": {"level": "floor", "empty": ["full",]},
        "password": "knockknock"
    }

    def setUp(self):
        self.original_conf_paths = copy(Configuration.CONF_PATHS)
        Configuration.CONF_PATHS = []

        self.config_file = tempfile.NamedTemporaryFile(suffix=".yaml", delete=False).name
        with open(self.config_file, "w") as conf:
            yaml.dump(self.FIXTURE, conf, default_flow_style=False)

        os.environ['TESTING_CONFIRE_PASSWORD'] = 'password'

    def tearDown(self):
        Configuration.CONF_PATHS = self.original_conf_paths
        os.remove(self.config_file)

        os.environ.pop('TESTING_CONFIRE_PASSWORD')
        assert 'TESTING_CONFIRE_PASSWORD' not in os.environ

    def test_search_path(self):
        """
        Assert there are directories to search for a configuration
        """
        self.assertGreater(len(self.original_conf_paths), 0)

    def test_empty_conf_path(self):
        """
        Test that an empty conf path raises no exceptions
        """
        config = TestConfiguration.load()
        self.assertEquals(len(Configuration.CONF_PATHS), 0)
        self.assertTrue(config['mysetting'])
        self.assertIsNone(config.get('myprop'))

    def test_load_config(self):
        """
        Assert config can load from YAML
        """
        Configuration.CONF_PATHS.append(self.config_file)

        config = TestConfiguration.load()
        self.assertIsNotNone(config.get('myprop'))
        self.assertEqual(config.get('myprop'), "Allen")
        self.assertTrue(isinstance(config.get('nested'), NestedConfiguration))
        self.assertEqual(config.get("nested").get("level"), "floor")

    def test_load_override(self):
        """
        Assert that loading config overrides default
        """
        Configuration.CONF_PATHS.append(self.config_file)

        config = TestConfiguration.load()
        self.assertFalse(config["mysetting"])

    def test_configure_by_dict(self):
        """
        Check configuration by dictionary
        """
        config = TestConfiguration.load()
        config.configure({"anoption":45, "foo":"bar"})
        self.assertEqual(config["anoption"], 45)
        self.assertEqual(config["foo"], "bar")

    def test_configure_by_conf(self):
        """
        Check configuration by other configuration
        """
        configa = TestConfiguration.load()
        configb = TestConfiguration.load()

        self.assertEqual(configa["anoption"], configb["anoption"])

        configa.anoption = 80
        self.assertNotEqual(configa["anoption"], configb["anoption"])

        configb.configure(configa)
        self.assertEqual(configa["anoption"], configb["anoption"])

    def test_configure_with_none(self):
        """
        Ensure None passed to configure doesn't break
        """
        config = TestConfiguration.load()
        try:
            config.configure(None)
        except Exception:
            self.fail("None passed to configure raised an error!")

    def test_nested_configure(self):
        """
        Ensure nested configurations work
        """
        config = TestConfiguration.load()
        data   = {"nested": {"nested": {"level":"basement"}, "level": "lobby"}}
        config.configure(data)
        self.assertEqual(type(config.get('nested')), NestedConfiguration)
        self.assertEqual(type(config.get('nested').get('nested')), SubNestedConfiguration)
        self.assertEqual(config.get('nested').get('level'), 'lobby')
        self.assertEqual(config.get('nested').get('nested').get('level'), 'basement')

    def test_environ_configuration(self):
        """
        Test the environ setting on a config
        """
        config = TestConfiguration.load()
        self.assertEqual('password', config.get('password'))

    def test_settings_file_environ_override(self):
        """
        Test that the settings file overrides the environ
        """
        Configuration.CONF_PATHS.append(self.config_file)

        config = TestConfiguration.load()
        self.assertEqual('knockknock', config.get('password'))

    def test_options(self):
        """
        Test the options method
        """
        config = TestConfiguration.load()
        options = dict(config.options())

        self.assertIn("mysetting", options)
        self.assertIn("anoption", options)
        self.assertIn("paththere", options)

        self.assertNotIn("_notanopt", options)
        self.assertNotIn("amethod", options)
        self.assertNotIn("NOTANOPT", options)

    def test_get(self):
        """
        Assert that get returns default or key
        """
        config = TestConfiguration.load()
        self.assertTrue(config.get("mysetting"))
        self.assertNotIn("notanopt", dict(config.options()))
        self.assertEqual(config.get("notanopt", 1), 1)

    def test__get__(self):
        """
        Check the getkey method
        """
        config = TestConfiguration.load()
        self.assertTrue(config["mysetting"])

    def test_case_insensitivity(self):
        """
        Assert case insensitivity in getitem
        """
        config = TestConfiguration.load()
        self.assertTrue(config["MYSETTING"])

    def test_key_error(self):
        """
        Assert not found key raises an exception
        """
        with self.assertRaises(KeyError):
            config = TestConfiguration.load()
            self.assertTrue(config["notanopt"])

    @unittest.skip
    def test_configure_back_to_empty(self):
        """
        Can override an empty list or dictionary from configuration
        """
        # Setup normal configuration
        Configuration.CONF_PATHS.append(self.config_file)
        config = TestConfiguration.load()
        self.assertGreater(len(config.nested.empty), 0, "Configuration not loaded")

        # Now reoverride with orignal settings - like the TestingConfig
        config.configure(NestedConfiguration())
        self.assertEqual(config.nested.level, 1)
        self.assertEqual(len(config.nested.empty), 0)

class SubNestedConfiguration(Configuration):

    level = 2

class NestedConfiguration(Configuration):

    level  = 1
    empty  = []
    nested = SubNestedConfiguration()

class TestConfiguration(Configuration):
    """
    A subclass of the Configuration class for testing purposes.
    """

    _notanopt = "joe"
    NOTANOPT  = "bob"
    mysetting = True
    anoption  = 42
    paththere = "/var/log/there.pth"
    nested    = NestedConfiguration()
    password  = environ_setting('TESTING_CONFIRE_PASSWORD')

    def amethod(self):
        return True


confire/config.py
# confire.conf
# A simple configuration module for Confire
#
# Author:   Benjamin Bengfort <ben@cobrain.com>
# Created:  Tue May 20 22:19:11 2014 -0400
#
# Copyright (C) 2013 Cobrain Company
# For license information, see LICENSE.txt
#
# ID: conf.py [] ben@cobrain.com $

"""
Confire class for specifying Confire specific optional items via a YAML
configuration file format. The main configuration class provides utilities
for loading the configuration from disk and iterating across all the
settings. Subclasses of the Configuration specify defaults that can be
updated via the configuration files.

General usage:

    from confire.conf import settings
    mysetting = settings.get('mysetting', default)

You can also get settings via a dictionary like access:

    mysetting = settings['mysetting']

However, this will raise an exception if the setting is not found.

Note: Keys are CASE insensitive

Note: Settings can be modified directly by settings.mysetting = newsetting
however, this is not recommended, and settings should be fetched via the
dictionary-like access.
"""

##########################################################################
## Imports
##########################################################################

import os
import yaml
import warnings

from copy import deepcopy
from .exceptions import ImproperlyConfigured, ConfigurationMissing

##########################################################################
## Environment helper function
##########################################################################

def environ_setting(name, default=None, required=True):
    """
    Fetch setting from the environment. The bahavior of the setting if it
    is not in environment is as follows:

        1. If it is required and the default is None, raise Exception
        2. If it is requried and a default exists, return default
        3. If it is not required and default is None, return  None
        4. If it is not required and default exists, return default
    """
    if name not in os.environ and default is None:
        message = "The {0} ENVVAR is not set.".format(name)
        if required:
            raise ImproperlyConfigured(message)
        else:
            warnings.warn(ConfigurationMissing(message))

    return os.environ.get(name, default)

##########################################################################
## Configuration Base Class
##########################################################################

class Configuration(object):
    """
    Base configuration class specifies how configurations should be
    handled and provides helper methods for iterating through options and
    configuring the base class.

    Subclasses should provide defaults for the various configurations as
    directly set class level properties. Note, however, that ANY directive
    set in a configuration file (whether or not it has a default) will be
    added to the configuration.

    Example:

        class MyConfig(Configuration):

            mysetting = True
            logpath   = "/var/log/myapp.log"
            appname   = "MyApp"

    The configuration is then loaded via the classmethod `load`:

        settings = MyConfig.load()

    Access to properties is done two ways:

        settings['mysetting']
        settings.get('mysetting', True)

    Note: None settings are not allowed!
    """

    CONF_PATHS = [
        '/etc/confire.yaml',                    # The global configuration
        os.path.expanduser('~/.confire.yaml'),  # User specific configuration
        os.path.abspath('conf/confire.yaml')    # Local directory configuration
    ]

    @classmethod
    def load(klass):
        """
        Insantiates the configuration by attempting to load the
        configuration from YAML files specified by the CONF_PATH module
        variable. This should be the main entry point for configuration.
        """
        config = klass()
        for path in klass.CONF_PATHS:
            if os.path.exists(path):
                with open(path, 'r') as conf:
                    config.configure(yaml.load(conf))
        return config

    def configure(self, conf={}):
        """
        Allows updating of the configuration via a dictionary of
        configuration terms or a configuration object. Generally speaking,
        this method is utilized to configure the object from a JSON or
        YAML parsing.
        """
        if not conf: return
        if isinstance(conf, Configuration):
            conf = dict(conf.options())
        for key, value in conf.items():
            opt = self.get(key, None)
            if isinstance(opt, Configuration):
                opt.configure(value)
            else:
                self.__dict__[key] = value

    def options(self):
        """
        Returns an iterable of sorted option names in order to loop
        through all the configuration directives specified in the class.
        """
        keys = self.__class__.__dict__.copy()
        keys.update(self.__dict__)
        keys = sorted(keys.keys())

        for opt in keys:
            val = self.get(opt)
            if val is not None:
                yield opt, val

    def get(self, key, default=None):
        """
        Fetches a key from the configuration without raising a KeyError
        exception if the key doesn't exist in the config, instead it
        returns the default (None).
        """
        try:
            return self[key]
        except KeyError:
            return default

    def __getitem__(self, key):
        """
        Main configuration access method. Performs a case insensitive
        lookup of the key on the class, filtering methods and pseudo
        private properties. Raises KeyError if not found. Note, this makes
        all properties that are uppercase invisible to the options.
        """
        key = key.lower()
        if hasattr(self, key):
            attr = getattr(self, key)
            if not callable(attr) and not key.startswith('_'):
                return attr
        raise KeyError("%s has no configuration '%s'" % (self.__class__.__name__, key))

    def __repr__(self):
        return str(self)

    def __str__(self):
        s = ""
        for opt, val in self.options():
            r = repr(val)
            r = " ".join(r.split())
            wlen = 76-max(len(opt),10)
            if len(r) > wlen:
                r = r[:wlen-3]+"..."
            s += "%-10s = %s\n" % (opt, r)
        return s[:-1]


confire/__init__.py
# confire
# A simple app configuration scheme using YAML and class based defaults.
#
# Author:   Benjamin Bengfort <benjamin@bengfort.com>
# Created:  Sun Jul 20 09:44:32 2014 -0400
#
# Copyright (C) 2014 Bengfort.com
# For license information, see LICENSE.txt
#
# ID: __init__.py [] benjamin@bengfort.com $

"""
A simple app configuration scheme using YAML and class based defaults.
"""

##########################################################################
## Imports
##########################################################################

from .config import Configuration, environ_setting
from .exceptions import ImproperlyConfigured

##########################################################################
## Package constants
##########################################################################

__version__ = (0, 2, 0)

##########################################################################
## Package helpers
##########################################################################

def get_version():
    """
    Returns a string containing the version number
    """
    return ".".join("%i" % idx for idx in __version__)


confire/exceptions.py
# confire.exceptions
# Exceptions hierarchy for Confire
#
# Author:   Benjamin Bengfort <benjamin@bengfort.com>
# Created:  Mon Jul 21 11:02:09 2014 -0400
#
# Copyright (C) 2014 Bengfort.com
# For license information, see LICENSE.txt
#
# ID: exceptions.py [] benjamin@bengfort.com $

"""
Exceptions hierarchy for Confire
"""

##########################################################################
## Exceptions Hierarchy
##########################################################################

class ConfireException(Exception):
    """
    Base class for configuration exceptions.
    """
    pass

class ImproperlyConfigured(ConfireException):
    """
    The user did not properly set a configuration value.
    """
    pass

##########################################################################
## Warnings Hierarchy
##########################################################################

class ConfireWarning(Warning):
    """
    Base class for configuration warnings.
    """
    pass

class ConfigurationMissing(ConfireWarning):
    """
    Warn the user that an optional configuration is missing.
    """
    pass


confire/example.py
# confire.example
# This is an example configuration file
#
# Author:   Benjamin Bengfort <benjamin@bengfort.com>
# Created:  Sun Jul 20 14:56:39 2014 -0400
#
# Copyright (C) 2014 Bengfort.com
# For license information, see LICENSE.txt
#
# ID: example.py [] benjamin@bengfort.com $

"""
This file is an example file that you might put into your code base to
have a configuration library at your fingertips!
"""

##########################################################################
## Imports
##########################################################################

import os
from confire import Configuration

##########################################################################
## Database Configuration
##########################################################################

class DatabaseConfiguration(Configuration):
    """
    This object contains the default connections to a Postgres Database.
    """

    host            = "localhost"
    port            = 5432
    database        = "app"
    user            = "postgres"
    password        = os.environ.get("DATABASE_PASSWORD")

##########################################################################
## Confire Configuration Defaults
##########################################################################

class ExampleConfiguration(Configuration):
    """
    This object contains an example configuration.

    debug: allow debug checking
    testing: are we in testing mode?
    """

    CONF_PATHS = [
        '/etc/myapp.yaml',                    # The global configuration
        os.path.expanduser('~/.myapp.yaml'),  # User specific configuration
        os.path.abspath('conf/myapp.yaml'),   # Local directory configuration
        os.path.abspath('conf/example-config.yaml')
    ]

    debug           = True
    testing         = False
    database        = DatabaseConfiguration()

##########################################################################
## Import this loaded Configuration
##########################################################################

settings = ExampleConfiguration.load()

if __name__ == '__main__':
    print settings
